torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1522,7 +1522,13 @@ void ProcessGroupNCCL::watchdogHandler() {
     for (auto it = workMetaList_.begin(); it != workMetaList_.end();
          /* no increment */) {
       auto& work = *it;
-      work.checkAndSetException();
+      // When terminateProcessGroup_ is true, communicators have already been
+      // aborted, So cannot check exception based on them. But watchdog needs to
+      // finish the check for the works that have already been enqueued to
+      // workMetaList_
+      if (!terminateProcessGroup_.load()) {
```

Comment:
@zdevito has learned that aborting a communicator causes the work object to 'complete'.  If we log this completion it can be a bit confusion.  We could handle this several ways

1) try to flush the work cleanup loop for some timeout before we actually do the aborts
1 seems not great imo, we don't really want to add more delays before doing the user-requested abort, nor can we be sure the delay would be fruitful (if collectives are hanging or taking longer than our timeout)

2) just do the abort and let the watchdog loop observe what it observes.  But add a log into the flight recorder for 'comm abort events', which should make it pretty clear from reading the flight log that any 'observed completions' after an abort are spurious
This ties into Zach's recent PR https://github.com/pytorch/pytorch/pull/119249

Reply:
Discussed offline that we will roll out the change with knobs and will leave Flight recorder change to another PR